Cryo Tube and Healing Changes

Aurum22 opened the pull request:
Fix #25169 

Cryotubes no longer do healing without a contained drug.

Buffs cryoxadone and clonexadone to be a bit more useful. 

Adds nanite fluid, which is like cryoxadone for mechanical parts. It can be made with cryoxadone, space lube, and aluminum chilled down nice and low (see cryoxadone). 

Peridaxon now uses cleaner code to do what it did before.

This is NOT safe to use on its own and should be used on conjunction with cryoxadone or clonexadone.

Supernatural effects that rely on heal_organ_damage will still include robotic parts.

<!-- 
Do not forget to add a changelog when you have made admin/player facing changes that can alter gameplay.
Examples which require a changelog entry include:
* Adding/removing objects that players may interact with, or the way they function.
* Adding/removing/altering admin tools.
* Changing the map.

Examples were changelog entries are optional/not typically required:
* Cosmetic changes such as descriptions, sound effects, etc.
* Optimizations and other changes to underlying systems which do not affect gameplay.
* Minor bug fixes.

You'll find a README and example file in .\html\changelogs\ for further instructions.

You can also find a template for adding your changelog directly to the PR description here: https://github.com/Baystation12/Baystation12/wiki/Automatic-changelog-generation
-->

🆑 Cronac
bugfix: Cryogenic tubes will no longer heal anything without containing a drug.
tweak: Clonexadone and Cryoxadone have been buffed due to cryogenic tube changes.
rscadd: Adds nanite fluid, which is like cryoxadone for robotic parts.  It can be made with cryoxadone, space lube, and aluminum chilled down to -25C. Requires 5 phoron as a catalyst. This is NOT safe to use on its own and should be used on conjunction with cryoxadone or clonexadone.
/🆑

afterthought2 commented on code/game/machinery/cryo.dm:
```diff
@@ -257,11 +257,11 @@
+				occupant.adjustToxLoss(max(-1, -10/occupant.getToxLoss(), 0))
 			var/heal_brute = occupant.getBruteLoss() ? min(1, 20/occupant.getBruteLoss()) : 0
 			var/heal_fire = occupant.getFireLoss	() ? min(1, 20/occupant.getFireLoss()) : 0
 			occupant.heal_organ_damage(heal_brute,heal_fire)
```
Just remove this whole block entirely; don't add arguments to `adjustToxLoss`.

Aurum22 replied:
Done, cryotube will no longer do any healing on its own.

afterthought2 commented on code/modules/mob/living/carbon/human/human_damage.dm:
```diff
@@ -215,8 +215,16 @@
+				amount = 0
+
+		else if(heal && robo)
+			if((I.damage < amount) && BP_IS_ROBOTIC(I))
```
Don't change this proc.

Aurum22 replied:
Undid this.

afterthought2 commented on code/modules/reagents/Chemistry-Reagents/Chemistry-Reagents-Medicine.dm:
```diff
@@ -172,8 +172,19 @@
+					H.drowsyness++
+					if(I.damage >= I.min_bruised_damage)
+						continue
+				I.damage = max(I.damage - (15*removed), 0)
```
Never modify the organ's damage directly; call the appropriate proc instead. Much more must be done internally by the organ. Don't do special brain handling here; don't mutate confused/drowsy without using proper proc calls on the mob.

Aurum22 replied:
A few pieces of code rely on this such as mineral baths and peridaxon. I fixed this and will tweak the mineral bath in my other PR.

afterthought2 commented on code/modules/reagents/Chemistry-Reagents/Chemistry-Reagents-Medicine.dm:
```diff
@@ -195,6 +206,39 @@
+					H.drowsyness++
+					if(I.damage >= I.min_bruised_damage)
+						continue
+				I.damage = max(I.damage - (25*removed), 0)
```
Combine this with the code for cryoxadone; it's the same with a var change.

Aurum22 replied:
Subtyping cryoxadone and clonexadone would be a lot of work (changing other files that utilize the chem etc) for pretty much no return seeing how I don't think we'll be adding any more cryogenic healing drugs any time soon.

afterthought2 commented on code/modules/reagents/Chemistry-Reagents/Chemistry-Reagents-Medicine.dm:
```diff
@@ -195,6 +206,39 @@
+		M.add_chemical_effect(CE_OXYGENATED, 1)
+		M.heal_organ_damage(15 * removed, 15 * removed, affect_robo = 1)
+		M.add_chemical_effect(CE_PULSE, -2)
+		M.adjustToxLoss((-10 * removed),1)
```
Don't have it do any of this stuff, just repair robo organs slowly.

Aurum22 replied:
Trimmed.

comma commented:
I think changelog should mention change to Peridaxon (assuming it's intended) - that it won't heal damage if organ is beyond bruised threshold

Aurum22 commented:
Good catch, this is only intended for the brain, which is how it currently works. I'll update that now.

comma commented on code/modules/reagents/Chemistry-Reagents/Chemistry-Reagents-Medicine.dm:
```diff
@@ -193,8 +199,32 @@
+			var/mob/living/carbon/human/H = M
+			for(var/obj/item/organ/internal/I in H.internal_organs)
+				if(BP_IS_ROBOTIC(I))
+					I.heal_damage(20*removed, 20*removed)
```
internal organs only take one arg to heal_damage
`I.heal_damage(20*removed)`

Aurum22 replied:
Fixed

comma commented on code/modules/reagents/Chemistry-Reagents/Chemistry-Reagents-Medicine.dm:
```diff
@@ -406,14 +436,13 @@
-					continue
-			I.damage = max(I.damage - removed, 0)
+			if(!BP_IS_ROBOTIC(I))
+				I.heal_damage(removed, removed)
```
`I.heal_damage(removed)`

Aurum22 replied:
Fixed

afterthought2 commented on code/modules/reagents/Chemistry-Reagents/Chemistry-Reagents-Medicine.dm:
```diff
@@ -406,14 +436,13 @@
+					H.confused++
+					H.drowsyness++
+					if(I.damage >= I.min_bruised_damage)
+						continue
```
What's the point of this check if we already healed it? I think your new logic is incorrect.

Aurum22 replied:
I just updated it, thanks.

afterthought2 commented on code/modules/reagents/Chemistry-Reagents/Chemistry-Reagents-Medicine.dm:
```diff
@@ -406,14 +436,16 @@
+						H.confused++
+						H.drowsyness++
+				else
+					I.heal_damage(removed)
```
Notice how you are doing this twice? Try
```
				if(I.organ_tag == BP_BRAIN)
					if(I.damage >= I.min_bruised_damage)
						continue
					H.confused++
					H.drowsyness++
				I.heal_damage(removed)
```
Else is not your friend.

Aurum22 replied:
This is more efficient. I'll try to avoid "else" in the future.

afterthought2 replied:
Nah, `else` is as useful as every logic keyword, just the above pattern of doing checks first and breaking/continueing/returning if failed before the actual action happens often leads to more easily readable code with fewer complicated logic branches.

Aurum22 committed:
Cryohealing Changes

Cryotubes no longer do healing without a contained drug.

Buffs cryoxadone and clonexadone to be a bit more useful.

Adds nanite fluid, which is like cryoxadone for mechanical parts. It can be made with cryoxadone, space lube, and aluminum chilled down nice and low (see cryoxadone).

Peridaxon now uses cleaner code to do what it did before.

This is NOT safe to use on its own and should be used on conjunction with cryoxadone or clonexadone.

Supernatural effects that rely on heal_organ_damage will still include robotic parts.
